### DISCUSSION TITLE
Custom Classifier on top of all existing species in BirdNET

### DISCUSSION THREAD

**Discussion by safahafiz** [birdnet-team/BirdNET-Analyzer#275] [General]
Can I build a classifier on top of the existing one for all the species using new data. In other words, I want to add new species to the list that BirdNET already has. However, the custom classifier only tests the input with the species in the custom classifier. 

Is it possible to do transfer learning to train BirdNET?

**Comment by Josef-Haupt** [comment]
Hey, you can indeed extend BirdNET with your own classes. If you're using the GUI check the "append" option under Model Save Mode. In the CLI the option is called --model_save_mode in train.py.
This mode outputs the BirdNET classification and concats them to your custom classifier, so you get one long vector, with your new classes at the end.
Does that answer your question?
*text copied from the issues*

**Reply by FlorianH27** [reply]
Can "append" only be used to add new species to the existing model, or can it also be used to improve the performance for already implemented species?

**Reply by Josef-Haupt** [reply]
The current implementation is "stupid". It let's you train whatever classes for your new classifier and just concats the resulting vector to the BirdNET output. So redundant classes are possible, you just have to use the correct one in the vector.

**Reply by w-out** [reply]
A related question on using the "append" option. Whenever I try to run my appended model/classifier on multiple files, I get an error in the GUI. The only way it will work is when I manually remove the double species names from the species list that is created during the training process. However, I am afraid that removing these double species names will influence my results, as I get approximately the same results with my appended classifier compared to BirdNET's own model.

**Reply by luigisa1** [reply]
> The current implementation is "stupid". It let's you train whatever classes for your new classifier and just concats the resulting vector to the BirdNET output. So redundant classes are possible, you just have to use the correct one in the vector.

So retraining a class directly is not possible at all?

**Reply by kahst** [reply]
No, not at the moment. The original ("old") classifier (and with those the weights of the model) will remain unchanged if "append" mode is chosen.

**Reply by luigisa1** [reply]
> No, not at the moment. The original ("old") classifier (and with those the weights of the model) will remain unchanged if "append" mode is chosen.

So, if I aim to improve the accuracy for some species with new data that I have collected through field trials and labelled by experts, what would be the best strategy to follow?

**Reply by kahst** [reply]
If you have training data for the species and also have some background/non-target sounds, you could train a new classifier, chose "append" and after analysis combine the scores for the original class and your improved class. Let's assume you want to improve Skylark detection, you train a Skylark classifier, append, and then you'll get two Skylark scores for each segment containing a Skylark (old classifier and appended classifier). Now, sometimes the original classifier might miss the birds, sometimes your classifier might miss the bird, but if you have a good scheme of post-processing ("both scores need to be above 0.5", "base score needs to be above 0.7, new score should be at least 0.25") you should be able to improve detection accuracy.

**Reply by tphakala** [reply]
This is exactly what I have done, and this method works very well.

**Reply by luigisa1** [reply]
> Si tiene datos de entrenamiento para la especie y también algunos sonidos de fondo o no objetivo, puede entrenar un nuevo clasificador, seleccionar "Añadir" y, tras el análisis, combinar las puntuaciones de la clase original y la clase mejorada. Supongamos que desea mejorar la detección de alondras: entrena un clasificador de alondras, añade las puntuaciones y obtendrá dos puntuaciones de alondras por cada segmento que contenga una alondra (clasificador antiguo y clasificador añadido). Ahora bien, a veces el clasificador original puede no detectar las aves, a veces el clasificador puede no detectarlas, pero si cuenta con un buen sistema de posprocesamiento ("ambas puntuaciones deben ser superiores a 0,5", "la puntuación base debe ser superior a 0,7, la nueva puntuación debe ser al menos 0,25"), debería poder mejorar la precisión de la detección.

Perfect. Thank you very much for your help
